text/0135-template-class-object-binding.md

```diff
@@ -0,0 +1,105 @@
+      this.variant ?? `slds-button_${this.variant}`,
+      this.position ?? `slds-button_${this.position}`,
+      {
+        "slds-button_full-width": this.fullWidth,
+        "slds-button_stretch": this.stretch,
+      },
+    ];
+  }
+}
+```
+
+```html
+<template>
+  <button class={className}>
+    <slot></slot>
+  </button>
+</template>
+```
+
+## Motivation
+
+Manipulating an element's class list is a very common use case for UI frameworks. Since `class` is a standard Element attribute, component authors currently bind those attributes with a string value. However, generating complex class names through string concatenation can be hard to read and error-prone. This is the reason why this RFC proposes enhancements to the element `class` attribute binding.
+
+Popular UI frameworks offer such capability out of the box:
+
+- _Vue_: Use the standard [`class`](https://vuejs.org/guide/essentials/class-and-style) HTML attribute as described in this proposal.
+- _Svelte_: Use a custom template directive called [`class:name`](https://svelte.dev/docs/element-directives#class-name), to conditionally apply classes.
+- _Angular_: Use a custom template directive called [`[ngClass]`](https://angular.io/api/common/NgClass), with a similar API as described in this proposal.
+
+It is also worth mentioning that React has no built-in support for conditional style application. However, most React applications use the [classnames](https://www.npmjs.com/package/classnames) NPM package to fill this gap (6.3M weekly downloads).
+
+The `classSet` utility method exposed from the Salesforce internal `lightning/utils` module offers similar capabilities as the public classnames NPM package. This utility function is used by +500 components in Salesforce's internal code base. As this LWC module isn't exposed on the platform, platform components have to re-implement such utility methods for managing complex styles.
+
+As a side note, modern UI frameworks also have built-in capabilities to manipulate element styles. This capability is out of the scope of this RFC and will be the subject of a follow-up proposal.
+
+## Detailed design
+
+This RFC proposes to change the semantics of the template `class` attribute. On top of accepting a string value, the `class` attribute would now accept objects (plain objects and arrays) representing the CSS classes applied to the elements. The LWC engine would then interpret the object to determine which class names should be applied to the element.
+
+The list of class names applied to a given element is defined as follows:
+
+1. If the value is a `string`, apply all the classes in the value.
+2. If the value is an `object`, iterate over its keys and apply the classes with the keys associated with a truthy value.
+3. If the value is an `array`, iterate over its entries and apply this algorithm recursively for each of its entries.
+4. Ignore all other values: `null`, `undefined`, other primitive values, and complex objects.
```

Comment:
This might be covered with the "complex objects" but wanted to ask, with template expressions component authors can use complex expressions such as arrow functions.

Is that something you think the engine should handle?

For example: 
```js
<button class={() => "slds-button__icon"}>
```

Reply:
Great question, your example is equivalent to:

```js
export default class extends LightningElement {
  className = () =>  "slds-button__icon";
}
```

```html
<button class={className}>
```

Since the object bound to the `class` attributes isn't a string, an array, nor a plain object the generated class name ``. In the current version of the engine, the class name will be `() => &quot;slds-button__icon&quot;`  (coercing the function object to string).